packages/solid-urql/src/createQuery.ts

```diff
@@ -0,0 +1,334 @@
+ * Given {@link CreateQueryArgs.query}, it executes the GraphQL query with the
+ * context’s {@link Client}.
+ *
+ * The returned result updates when the `Client` has new results
+ * for the query, and changes when your input `args` change.
+ *
+ * Additionally, if the `suspense` option is enabled on the `Client`,
+ * the `createQuery` hook will suspend instead of indicating that it’s
+ * waiting for a result via {@link CreateQueryState.fetching}.
+ *
+ * @example
+ * ```tsx
+ * import { gql, createQuery } from '@urql/solid';
+ *
+ * const TodosQuery = gql`
+ *   query { todos { id, title } }
+ * `;
+ *
+ * const Todos = () => {
+ *   const [result, reExecuteQuery] = createQuery({
+ *     query: TodosQuery,
+ *   });
+ *   // ...
+ * };
+ * ```
+ */
+export const createQuery = <
+  Data = any,
+  Variables extends AnyVariables = AnyVariables
+>(
+  args: CreateQueryArgs<Data, Variables>
+): CreateQueryResult => {
+  const client = useClient();
+  const getContext = asAccessor(args.context),
+    getPause = asAccessor(args.pause),
+    getRequestPolicy = asAccessor(args.requestPolicy),
+    getVariables = asAccessor(args.variables);
+
+  const resultSourceSubject = makeSubject<
+    OperationResultSource<OperationResult<Data, Variables>> | undefined
+  >();
+
+  // Combine suspense param coming from context and client with context being priority
+  const isSuspense = createMemo(() => {
+    const ctx = getContext();
+    if (ctx !== undefined && ctx.suspense !== undefined) {
+      return ctx.suspense;
+    }
+
+    return client.suspense;
+  });
+
+  const request = createRequest(args.query, getVariables() as any);
+  const context: Partial<OperationContext> = {
+    requestPolicy: getRequestPolicy(),
+    ...getContext(),
+  };
+  const operation = client.createRequestOperation('query', request, context);
+  const initialResult: CreateQueryState<Data, Variables> = {
+    operation: operation,
+    fetching: false,
+    data: undefined,
+    error: undefined,
+    extensions: undefined,
+    hasNext: false,
+    stale: false,
+  };
+
+  const [result, setResult] =
+    createStore<CreateQueryState<Data, Variables>>(initialResult);
+
+  // `createResource` is only way to trigger suspense, suspense handling works as follows:
+  // 1. Calling `dataResource` mounts resource to closes suspense. Thus depending on where it is called
+  // it could mount in same component or parent. For usage convince we call it by proxy when using `data`
+  // 2. After any changes we call `refetch` which results in creation of new promise and depending on `fetching`
+  // value we either discard promise without resolving or resolve it. Discard doesn't result in anything while
+  // resolve finishes suspense
+  const [dataResource, { refetch, mutate }] = createResource<
+    CreateQueryState<Data, Variables>
+  >(() => {
+    return new Promise(resolve => {
+      const got = unwrap(result);
+      if (got.fetching) {
+        return;
+      }
+
+      resolve(got);
```

Comment:
I don't quite feel this will be the best approach to be honest. Especially, since the inputs can change while we're waiting to resolve a promise.

I also haven't checked this in detail, but I'm very hesitant to use this approach of resolving + re-triggering this as per the `refetch` call. This seems a little brittle to me.

You can see how the `vue` bindings do this: https://github.com/urql-graphql/urql/blob/344b54410aaaf3927c70174ad8e810ccceff336d/packages/vue-urql/src/useQuery.ts#L389-L403

The only difference is that we wouldn't necessarily want to check for `stale`. That depends on how background fetches should be handled.

Maybe it does make sense though to include this, since not using suspense would be the escape hatch here. But since in these bindings there's no concept of conditionally using the resource, it's likely better not to check for `stale`

The important part here is that the promise _actually_ only reacts to the result. While the `source` isn't shared in the Vue implementation and instead relies on the semaphore logic in the `Client`, it's also fine to try to share it

Reply:
Tried to implement it in similar way as in vue, technically it seems to be working (you can take a look at https://github.com/urql-graphql/urql/pull/3327/commits/b0725fea415f60a49ccf22720dbfa7004404820e). Thought after working a bit on it I'm not sure if this is good idea as it results in dependency on subscription order as subscription to fetch data has to be made before subscription in `createResource` fetcher.

Furthermore I'm not 100% sure if there wont be cases where order is different

Reply:
Order should be fine after some testing I don't think there should be any edge cases